### PR TITLE
GPII-3572/GPII-4244: Metrics switch, and site-config updates

### PR DESCRIPTION
**Pull request by stegru**
Makes the service automatically updates the site-config file (GPII-4244). It also reads the `metricsSwitch` value from the siteConfig and starts Morphic with the appropriate `NODE_ENV` environment variable.

Configurable here (disabled by default): https://github.com/GPII/gpii-app/blob/2e99c3f5db3bcd236737b623e287361049d9029c/provisioning/service.json5#L18

Works with https://github.com/GPII/gpii-app/pull/73

This includes the GPII-2338 branch (https://github.com/GPII/windows/pull/155), the changes related to this PR are https://github.com/stegru/windows/compare/GPII-2338...stegru:GPII-4244?expand=1

### CODE REVIEW

File: gpii-service/src/processHandling.js [hunk 1]
```diff
@@ -113,11 +115,54 @@ processHandling.getProcessList = function (allProcesses) {
...
+
+        service.logDebug("Setting desktop icon: ", desktop, stashed, morphicHide ? "hide" : "show");
+
+        try {
+            // Ensure there's a copy in the stash location
+            if (!fs.existsSync(stashed)) {
+                fs.copyFileSync(desktop, stashed);
+            }
+
+            if (morphicHide) {
```
Comment (javihernandez) [inline gpii-service/src/processHandling.js]: I guess we also need a "morphicShow" - this is, to ensure that the shortcuts are copied back to the desktop when we switch from off to on.

Comment (stegru) [reply gpii-service/src/processHandling.js]: oops

**Comment by javihernandez** [comment]
ok to test

File: gpii-service/package.json [hunk 2]
```diff
@@ -19,11 +19,14 @@
     "dependencies": {
         "ffi-napi": "2.4.4",
         "json5": "2.1.0",
+        "mkdirp": "0.5.1",
         "minimist": "1.2.0",
-        "@gpii/os-service": "2.2.0",
+        "@gpii/os-service": "stegru/node-os-service#GPII-3934",
```
[the remark this answers is not in the document]
Comment (stegru) [reply gpii-service/package.json]: That change wasn't intended for this branch - reverted it.

File: gpii-service/src/service.js [hunk 3]
```diff
@@ -59,6 +59,31 @@ service.logImportant = logging.important;
...
 
+/**
+ * @typedef {Object} Config
+ * @property {Object<String,ProcessConfig>} processes Child process.
+ * @property {Object} logging Logging settings
+ * @property {String} secretFile The file containing site-specific information.
+ * @property {String} package.json The package.json file for gpii-app.
+ * @property {String} siteConfigFile The site-config file.
+ * @property {AutoUpdateConfig} autoUpdate Auto update settings.
+ * @property {String} morphicVersion [generated] The version field from gpii-app's package.json.
```
Comment (amb26) [inline gpii-service/src/service.js:70]: @javihernandez - looks like we are not actually publishing these to npm via fluid-publish or whatever as part of our release process - https://www.npmjs.com/package/gpii-app - I assume we need to actually do this before this facility becomes useful?

File: gpii-service/README.md [hunk 4]
```diff
@@ -102,7 +102,68 @@ This gets installed in `c:\Program Files (x86)\Morphic\windows\service.json5`.
...
-    "secretFile": "%ProgramData%\\Morphic Credentials\\secret.txt"
+    "secretFile": "%ProgramData%\\Morphic Credentials\\secret.txt",
+    // The gpii-app package.json file (default: "resources/app/package.json")
+    "package.json": "resources/app/package.json",
+
+    // Auto update of files
+    "autoUpdate": {
+        "enabled": false, // true to enable
+        // Where to store the 'last update' info
+        "lastUpdatesFile": "%ProgramData%\\Morphic\\last-updates.json5",
```
Comment (amb26) [inline gpii-service/README.md:113]: Shouldn't this more reasonably go in the morphic settings directory? Is there an expander syntax that will allow that?

File: gpii-service/src/service.js (hunk 3 above)
Comment (javihernandez) [reply gpii-service/src/service.js:70]: Actually, the machinery involved is:

1. Whatever is in the version field of the gpii-app's package.json file. No matter whether is published in npm or not.
2. https://github.com/GPII/gpii-morphic-site-resources, and when a new release is out, we must update the defaults (and any other site-specific defaults) according to the version that we have just released.

This is the approach we took when designing this, but given that there are no more pilot sites, the autoUpdate policy was turned on by default (see https://github.com/GPII/gpii-app/blob/master/provisioning/service.json5#L20). If by any chance we want to make this functionality work, we'd need to update the site-resources repo as indicated above and ship a service.json5 file with the autoUpdate functionality enabled.

Comment (amb26) [reply gpii-service/src/service.js:70]: OK - but I'm assuming that this pull is not part of your "secret sauce" 1.3.x branches so far, right? Despite it being marked as "@@@@" in the pad, when you say, "If by any chance we want to make this work" it sounds like it is not actually in the high priority list for delivered functionality at the very least for the case studies - does it sound like we want to turn this off by default before we merge this pull, just to make sure it doesn't destabilise anything and we can opt into it later if necessary?

Comment (javihernandez) [reply gpii-service/src/service.js:70]: IT IS part of 1.3.x, but we're not using the _autoUpdate_ functionality right now, it's disabled [by default](https://github.com/GPII/gpii-app/blob/master/provisioning/service.json5#L20), so we're fine to go. Note that we've used the 4 state metrics switch functionality from this PR, which "I think" is what was considered as a priority ("@@@@" in the pad).

**Pull request merged**